lnwire/custom_records.go

```diff
@@ -150,27 +145,104 @@ func (c CustomRecords) RecordProducers() []tlv.RecordProducer {
 	// Convert the custom records map to a TLV record producer slice.
 	records := tlv.MapToRecords(c)
 
-	// Convert the records to record producers.
-	producers := make([]tlv.RecordProducer, len(records))
-	for i, record := range records {
-		producers[i] = &recordProducer{record}
-	}
-
-	return producers
+	return RecordsAsProducers(records)
 }
 
 // Serialize serializes the custom records into a byte slice.
 func (c CustomRecords) Serialize() ([]byte, error) {
 	records := tlv.MapToRecords(c)
-	stream, err := tlv.NewStream(records...)
+	return EncodeRecords(records)
+}
+
+// SerializeTo serializes the custom records into the given writer.
+func (c CustomRecords) SerializeTo(w io.Writer) error {
+	records := tlv.MapToRecords(c)
+	return EncodeRecordsTo(w, records)
+}
+
+// ProduceRecordsSorted converts a slice of record producers into a slice of
+// records and then sorts it by type.
+func ProduceRecordsSorted(recordProducers ...tlv.RecordProducer) []tlv.Record {
+	records := fn.Map(func(producer tlv.RecordProducer) tlv.Record {
+		return producer.Record()
+	}, recordProducers)
+
+	// Ensure that the set of records are sorted before we attempt to
+	// decode from the stream, to ensure they're canonical.
+	tlv.SortRecords(records)
+
+	return records
+}
+
+// SortProducers sorts the given record producers by their type.
+func SortProducers(producers []tlv.RecordProducer) {
+	sort.Slice(producers, func(i, j int) bool {
+		recordI := producers[i].Record()
+		recordJ := producers[j].Record()
+		return recordI.Type() < recordJ.Type()
+	})
+}
+
+// TlvMapToRecords converts a TLV map into a slice of records.
+func TlvMapToRecords(tlvMap tlv.TypeMap) []tlv.Record {
+	tlvMapGeneric := make(map[uint64][]byte)
+	for k, v := range tlvMap {
+		tlvMapGeneric[uint64(k)] = v
+	}
+
+	return tlv.MapToRecords(tlvMapGeneric)
+}
+
+// RecordsAsProducers converts a slice of records into a slice of record.
+func RecordsAsProducers(records []tlv.Record) []tlv.RecordProducer {
+	return fn.Map(func(record tlv.Record) tlv.RecordProducer {
+		return &recordProducer{record}
+	}, records)
+}
```

Comment:
Ah yes, you're right... Though the `Record()` method on `tlv.Record` has pointer access, and we're using non-pointer slices of records. So we still have to do:
```
	return fn.Map(func(record tlv.Record) tlv.RecordProducer {
		return &record
	}, records)
```

I guess a nice helper function for the `fn` package would be `Ptr` which returns the pointer of a type.
With that it could just be `fn.Map(fn.Ptr, records)`.